infra/dlv/dlv-ctl.sh

```diff
@@ -0,0 +1,120 @@
+#!/bin/bash
+# Copyright 2016-2018 VMware, Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.!/bin/bash
+
+SSH="ssh -o StrictHostKeyChecking=no"
+SCP="scp -o StrictHostKeyChecking=no"
+
+REMOTE_DLV_ATTACH=/usr/local/bin/dlv-attach-headless.sh
+REMOTE_DLV_DETACH=/usr/local/bin/dlv-detach-headless.sh
+
+function usage() {
+    echo "Usage: $0 -h vch-address [-a/-d] [attach/detach] target" >&2
+    echo "Valid targets are: "
+    echo "    vic-init"
+    echo "    vic-admin"
+    echo "    docker-engine"
+    echo "    port-layer"
+    echo "    vic-machine"
+    exit 1
+}
+
+while getopts "h:ad" flag
+do
+    case $flag in
+
+        h)
+            # Optional
+            export DLV_TARGET_HOST="$OPTARG"
+            ;;
+
+        a)
+            export COMMAND="attach"
+            ;;
+
+        d)
+            export COMMAND="detach"
+            ;;
+
+        *)
+            usage
+            ;;
+    esac
+done
+
+shift $((OPTIND-1))
+
+if [ -n "${DOCKER_HOST}" -a -z "${DLV_TARGET_HOST}" ]; then
+    export DLV_TARGET_HOST=$(echo $DOCKER_HOST | cut -d ':' -f 1)
+fi
+
+if [ -z "${DLV_TARGET_HOST}" ]; then
+    usage
+fi
+
+if [[ -z "${COMMAND}" &&  $# != 2 ]]; then
+    usage
+elif [[ -n "${COMMAND}" && $# != 1 ]]; then
+    usage
+fi
+
+if [ -z "${COMMAND}" ]; then
+    COMMAND=$1
+    TARGET=$2
+else
+    TARGET=$1
+fi
+
+case ${TARGET} in
+
+    vic-init)
+        PORT=2345
+        ;;
+
+    vic-admin)
+        # Change target to vicadmin
+        TARGET=vicadmin
+        PORT=2346
+        ;;
+
+    docker-engine)
+        PORT=2347
+        ;;
+
+    port-layer)
+        PORT=2348
+        ;;
+
+    vic-machine)
+        PORT=2349
+        ;;
+
+    *)
+        usage
+        ;;
+esac
+
+if [ -z "${DLV_TARGET_HOST}" ]; then
+    usage
+fi
+
+if [ ${COMMAND} == "attach" ]; then
+    ${SSH} root@${DLV_TARGET_HOST} "nohup /usr/local/bin/dlv-attach-headless.sh $TARGET $PORT >  /var/tmp/${TARGET}.log 2>&1 &"
```

Comment:
The way this command is launched, there is no indication of failure if it occurs. I would suggest altering the command to `${SSH} root@${DLV_TARGET_HOST} "nohup /usr/local/bin/dlv-attach-headless.sh $TARGET $PORT >  /var/tmp/${TARGET}.log 2>&1 || cat /var/tmp/docker-engine.log'` so that in the event it fails, the user will see the reason without having to ssh in and check the log. It would be preferable to exit with an nonzero status in that case, also, so that this command can be properly used in other scripts and subshells. (For instance I don't use goland, so calling `dlv connect $(dlv-ctl attach args)` is a potential use-case for me since the last step prints out the IP to connect to)

Reply:
The reason I have it in the background is because ```dlv-ctl.sh attach``` does not behave like a real ```dlv attach```. In the latter case, killing the dlv process automatically detaches from the target. In case of dlv-ctl.sh you must use ```dlv-ctl.sh detach``` to perform the detach. However I agree that ssh'ing into the VCH is a nuisance. I am thinking of implementing a ```dlv-ctl.sh log``` command that dumps the content of the log.